File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"


using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), num_(0)
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10,
            std::bind(&NumberCounterNode::callbackNumber, this, std::placeholders::_1));
        

        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);


        server_ = this->create_service<example_interfaces::srv::SetBool>(
            "reset_counter",
            std::bind(&NumberCounterNode::callbackResetNumber, this, _1, _2));

        RCLCPP_INFO(this->get_logger(), "Number Counter has been started.");
    }

private:
    void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        num_ += msg->data;
        // RCLCPP_INFO(this->get_logger(), "%d", num_);
        auto msgOut = example_interfaces::msg::Int64();
        msgOut.data = num_;
        publisher_->publish(msgOut);
    }

    void callbackResetNumber(
        const example_interfaces::srv::SetBool::Request::SharedPtr request,
        const example_interfaces::srv::SetBool::Response::SharedPtr response
    )
    {   

        try
        {
            if (request->data)
            {
                num_ = 0;
                RCLCPP_INFO(this->get_logger(), "Counter has been set to zero");
            }        
            else
            {
                RCLCPP_INFO(this->get_logger(), "Counter was not reset");
            }
            response->success = true;
            response->message = "reset counter ran successfully";
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
            response->success = false;
            response->message = "Service call failed";
        }
    }

    int64_t num_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.num_ = 0
        self.subscriber_ = self.create_subscription(
            Int64, 'number', self.callback_number_counter, 10)
        self.server_ = self.create_service(
            SetBool, 'reset_counter', self.callback_reset_counter)
        self.get_logger().info('Number counter has been started')

    # Good naming standard to follow for subscribers
    #   callback_<topic name>()
    def callback_number_counter(self, msg):
        self.num_ += msg.data
        self.get_logger().info(str(self.num_))

    def callback_reset_counter(self, request, response):
        if request.data:
            self.num_ = 0
            response.success = True
            response.message = 'Counter has been reset'
        else:
            response.success = False
            response.message = 'Counter has not been reset'
        return response


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
